doc/cli_wrapper/shd-cli_wrapper_hld.md

```diff
@@ -0,0 +1,213 @@
+
+![Architecture](images/architecture_2.jpg)
+
+* CLI_wrapper server
+
+  CLI_wrapper server is responsible for the execution of click command. It starts when system boots up, and loads all needed libraries to server. Then setup a socket to wait client's request. Once it accepts a request, it spawns a CLI_worker process to deal with this request. 
+  
+  In order to avoid that some command may stuck in CLI worker and cause resource leakage, a watchdog thread is designed to watch the execution time of CLI worker, once the execution time exceeds maximum execution time, the CLI worker will be gracefully terminated. But if the CLI worker is still alive after one minute, a kill signal will be sent to CLI worker to tear down it. 
+
+  The local DB is used to store the record of spawned CLI worker, it records the steady timestamp and pid of the CLI worker. 
+
+* CLI_worker
+
+  Decode the request and execute this command. Return the execution result to client and finish this process.
+
+  Everytime cli_wapper server receives a request, it spawns a process to handle this request, so no single block point may happen in this design. 
+
+* CLI_wrapper client
+
+  It is a library that can be used to interact with CLI wrapper server. 
+
+---
+
+## 1.3 Design
+
+This section provides design details for CLI_wrapper.
+
+### 1.3.1 Generate wrapped files
+
+![Generate wrapper file](images/generate_wrapper_file.jpg)
+
+* In the step 1 and step 2, builder copies the necessary libraries to wrapped file and discard other libraries to reduce loading time
+
+* In the step 3, builder removes the content of command function body and replaces it with CLI_wrapper client
+
+### 1.3.2 Command execution flow
+
+![Command execution flow](images/command_execution_flow.jpg)
+
+* System ctrl bring up CLI_wrapper server when system starts
+
+* Click engine imports wrapped main.py and call CLI_wrapper client to send request
+
+* CLI_wrapper client gets command parameter and environment setting from CLI context, then encodes those information to a command string and send to CLI_wrapper server.
+
+* CLI_wrapper server launches a CLI worker to deal with this request and write timestamp and pid to local DB
+
+* CLI worker decodes the command to get parameters and environment setting, then execute this command based on those settings. In the end, it sends return code and data back to CLI wrapper client
+
+### 1.3.3 Watchdog execution flow
+

```

Comment:
is it possible to accept optional value for each command for execution time? sometimes, it's hard to predict the execution time of the command.